### PR TITLE
Branch monster factory BUO update

### PR DESCRIPTION
**Pull request by danwalkowski**
I forgot to mention:
 • the navigationController stack is much simpler
 • the project uses the Branch cocoapod, instead of directly including all the sources
 • there are about 100 fewer lines of code

downsides:
 • currently still uses original (four) embedded sharing libraries; BranchInvite pod seems to have some sharing support, but it does not build correctly.  What is the best sharing scheme?

### CODE REVIEW

File: BranchMonsterFactory/AppDelegate.h [hunk 1]
```diff
@@ -7,11 +7,14 @@
 //
 
 #import <UIKit/UIKit.h>
+#import "BranchUniversalObject.h"
 
 @interface AppDelegate : UIResponder <UIApplicationDelegate>
 
 @property (strong, nonatomic) UIWindow *window;
 
+@property BranchUniversalObject *initialMonsterOrNULL;
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.h]: interesting. so we killed the old place this was stored?

File: BranchMonsterFactory/AppDelegate.m [hunk 2]
```diff
@@ -8,65 +8,36 @@
...
 @implementation AppDelegate
-            
+
 
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
     
     // Initalize Branch and register the deep link handler
     // The deep link handler is called on every install/open to tell you if the user had just clicked a deep link
     Branch *branch = [Branch getInstance];
-    [branch initSessionWithLaunchOptions:launchOptions andRegisterDeepLinkHandler:^(NSDictionary *params, NSError *error){
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: so glad this is gone

File: BranchMonsterFactory/AppDelegate.m [hunk 3]
```diff
@@ -79,23 +50,23 @@ - (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceAppl
 }
 
 - (void)applicationWillResignActive:(UIApplication *)application {
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: can we just delete all these empty methods?

File: BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h [hunk 4]
```diff
@@ -1,34 +0,0 @@
-//
```
Comment (derrickstaten) [inline BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h]: so we're removing the SDK in favor of the cocoapod? I'm certainly glad. Did you commit the change removing this `ExternalLibs` folder? git on my machine now says ExternalLibs is not staged for changing, I'll have to troubleshoot that separately.

Comment (derrickstaten) [inline BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h]: Also if this is now Cocoapods, should we include the .xcworkspace file or is that just going to confuse users?

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m [hunk 5]
```diff
@@ -51,9 +53,18 @@ - (void)viewDidLoad {
         [currView.layer setCornerRadius:currView.frame.size.width/2];
     }
     
+    //if it's null, create a new default monster
+    if (!self.editingMonster) {
+        self.editingMonster = [[BranchUniversalObject alloc] initWithCanonicalIdentifier:[NSString stringWithFormat:@"MONSTER_%u", arc4random_uniform(10000)]];
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: instead of a random number, can we use the face index, the body index, the color index, and the monster name (and any other important variables)? that way it's truly a canonical identifier.

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m [hunk 6]
```diff
@@ -43,7 +45,7 @@ - (void)viewDidLoad {
     for (int i = 0; i < [self.colorViews count]; i++) {
         UIView *currView = [self.colorViews objectAtIndex:i];
         [currView setBackgroundColor:[MonsterPartsFactory colorForIndex:i]];
-        if (i == [MonsterPreferences getColorIndex])
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: so glad you killed MonsterPreferences

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m (hunk 5 above)
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: following up, perhaps we could just call `[[BranchUniversalObject alloc] init];` here and set a canonical identifier before generating a link

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m [hunk 7]
```diff
@@ -63,16 +74,16 @@ - (void)viewDidLoad {
     self.botViewLayerThree.dataSource = self;
     [self.botViewLayerThree registerClass:[ImageCollectionViewCell class] forCellWithReuseIdentifier:@"cell"];
     
-    [self.etxtName setText:[MonsterPreferences getMonsterName]];
+    [self.etxtName setText:[self.editingMonster getMonsterName]];
     
-    UIBarButtonItem *barButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self.etxtName action:@selector(resignFirstResponder)];
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: if you're removing code, go ahead and delete it completely

Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: can you remove everything that you commented out? makes for cleaner diffs. we don't usually keep code around, we can always locate it on github if we really need to

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m [hunk 8]
```diff
@@ -7,13 +7,15 @@
...
 #import "MonsterPartsFactory.h"
 #import "ImageCollectionViewCell.h"
+#import "MonsterViewerViewController.h"
+#import "BranchUniversalObject+MonsterHelpers.h"
 #import "Branch.h"
 
 @interface MonsterCreatorViewController () <UICollectionViewDataSource, UICollectionViewDelegate>
 
+
 @property (weak, nonatomic) IBOutlet UITextField *etxtName;
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: can we rename this to `nameTextField` or something like that?

File: BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m [hunk 9]
```diff
@@ -35,7 +39,6 @@ @interface MonsterViewerViewController () <MFMailComposeViewControllerDelegate,
 @property (weak, nonatomic) IBOutlet UIButton *cmdTwitter;
 @property (weak, nonatomic) IBOutlet UIButton *cmdFacebook;
 
-@property (weak, nonatomic) IBOutlet UITextView *etxtUrl;
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m]: so we pulled this completely? I think we still want to show a link in the app, in addition to the "Copy Share URL" button. That way people can see the link as soon as it's generated :)

**Comment by danwalkowski** [comment]
As of the news this morning, and the Universal Link changes from Apple, Branchsters should also be updated with Universal LInks ASAP.
I can do that, but I'd like to get this PR in first.

File: BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h (hunk 4 above)
Comment (danwalkowski) [inline BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h]: It seems typical to remove the workspace and extra files, but leave the podfile.
That way users can install the cocoapods themselves.

Comment (danwalkowski) [inline BranchMonsterFactory/ExternalLibs/Branch-SDK/BNCConfig.h]: I do not remember doing anything to remove the ExternalLibs stuff. I don't know what that is. It's hluge though.

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m (hunk 6 above)
Comment (danwalkowski) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: me too

File: BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m (hunk 9 above)
Comment (danwalkowski) [inline BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m]: ok sure, that will only take a couple minutes

**Comment by derrickstaten** [comment]
there's still a .xcworkspace file, but when you try to open it, it's blank. if we're removing it, we should remove it completely

File: BranchMonsterFactory/AppDelegate.m [hunk 10]
```diff
@@ -8,94 +8,128 @@
...
-        if ([params objectForKey:@"monster"]) {
-            [MonsterPreferences setMonsterName:[params objectForKey:@"monster_name"]];
-            [MonsterPreferences setFaceIndex:[[params objectForKey:@"face_index"] intValue]];
-            [MonsterPreferences setBodyIndex:[[params objectForKey:@"body_index"] intValue]];
-            [MonsterPreferences setColorIndex:[[params objectForKey:@"color_index"] intValue]];
+        if (BUO && [BUO.metadata objectForKey:@"monster"]) {
             
-            // Choose the monster viewer as the next view controller
-            nextVC = [storyboard instantiateViewControllerWithIdentifier:@"MonsterViewerViewController"];
+            BranchUniversalObject* receivedMonster = BUO;
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: can we move the asterisk over to be consistent with the rest of the code?

File: BranchMonsterFactory/AppDelegate.m [hunk 11]
```diff
@@ -8,94 +8,128 @@
...
-    [branch initSessionWithLaunchOptions:launchOptions andRegisterDeepLinkHandler:^(NSDictionary *params, NSError *error){
-        
-        UINavigationController *navController = (UINavigationController *)self.window.rootViewController;
-        NSString * storyboardName = @"Main";
-        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:storyboardName bundle: nil];
-        UIViewController *nextVC;
+    
+    //[branch setDebug];
+    
+    [branch initSessionWithLaunchOptions:launchOptions andRegisterDeepLinkHandlerUsingBranchUniversalObject:^(BranchUniversalObject *BUO, BranchLinkProperties *linkProperties, NSError * error) {
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: change `* error` to `*error`

File: BranchMonsterFactory/AppDelegate.m [hunk 12]
```diff
@@ -8,94 +8,128 @@
...
-            
-            // If a name has been saved in preferences, then this user has already created a monster
-            // Load the viewer
-            if (![MonsterPreferences getMonsterName]) {
-                [MonsterPreferences setMonsterName:@""];
+            if (self.justLaunched) {
                 
-                // Choose the monster viewer as the next view controller
-                nextVC = [storyboard instantiateViewControllerWithIdentifier:@"MonsterCreatorViewController"];
+                self.justLaunched = NO; //turn off
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: can we remove the comment? it's not adding much

File: BranchMonsterFactory/AppDelegate.m [hunk 13]
```diff
@@ -8,94 +8,128 @@
...
+                }
+            } else if(self.foregrounded) {
+                if (receivedMonster) {
+                    self.initialMonster = receivedMonster;
+                    self.foregrounded = NO;
+                    [[NSNotificationCenter defaultCenter] postNotificationName:@"pushEditAndViewerViews" object:nil];
+                }
             }
+            
+        } else {
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: the effect of this `else` block is to push new view controllers, wiping out the monster i was working on, even if i just backgrounded the app for a second. I think we should re-think this

File: BranchMonsterFactory/AppDelegate.m [hunk 14]
```diff
@@ -8,94 +8,128 @@
...
-        // launch the next view controller
-        [navController setViewControllers:@[nextVC] animated:YES];
+    
     }];
     
     return YES;
 }
 
+- (BranchUniversalObject *) emptyMonster {
+
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: let's remove the extra spacing at the top of this method

File: BranchMonsterFactory/AppDelegate.m [hunk 15]
```diff
@@ -8,94 +8,128 @@
...
+- (void)applicationWillResignActive:(UIApplication *)application {
+}
+
+- (void)applicationDidEnterBackground:(UIApplication *)application {
+    
 }
 
 - (void)applicationWillEnterForeground:(UIApplication *)application {
-
+    NSLog(@"foreground");
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: please remove

File: BranchMonsterFactory/AppDelegate.m [hunk 16]
```diff
@@ -8,94 +8,128 @@
...
+    self.foregrounded = YES;
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: not sure what this flag is for, but it worries me that we're using a boolean flag here at all. It seems like this is supposed to prevent the issue I noted above (https://github.com/BranchMetrics/Branch-Example-Deep-Linking-Branchster-iOS/pull/5/files#r47735069), but it unfortunately doesn't

File: BranchMonsterFactory/AppDelegate.m [hunk 17]
```diff
@@ -8,94 +8,128 @@
...
-            } else {
-                nextVC = [storyboard instantiateViewControllerWithIdentifier:@"MonsterViewerViewController"];
+                if (receivedMonster) {
+                    self.initialMonster = receivedMonster;
+                    [[NSNotificationCenter defaultCenter] postNotificationName:@"pushEditAndViewerViews" object:nil];
+                } else {
+                    self.initialMonster = [self emptyMonster];
+                    [[NSNotificationCenter defaultCenter] postNotificationName:@"pushEditView" object:nil];
+                }
+            } else if(self.foregrounded) {
```
Comment (derrickstaten) [inline BranchMonsterFactory/AppDelegate.m]: please add a space between `if` and `(`

**Comment by derrickstaten** [comment]
this may not be a new issue, but for some reason i see a light outline to the top and left of the monster, like a subview was shifted slightly:
slack_for_ios_upload_720

**Comment by derrickstaten** [comment]
Love the Universal Link working :)

**Comment by derrickstaten** [comment]
Seems like the Facebook functionality may be broken somehow. It works on the live App Store version

**Comment by danwalkowski** [comment]
crap, that means it's an old library.  We can either disable bitcode or get rid of the 4 share buttons and bring up the BUO share sheet.  I'll try the share sheet and see if it works any better.

**Comment by danwalkowski** [comment]
Oh, the tiny outline happened after opening the project in Xode 7.2.  I'll have a look at it.

File: BranchMonsterFactory/AppDelegate.m (hunk 10 above)
Comment (danwalkowski) [inline BranchMonsterFactory/AppDelegate.m]: sure

File: BranchMonsterFactory/AppDelegate.m (hunk 11 above)
Comment (danwalkowski) [inline BranchMonsterFactory/AppDelegate.m]: sure

File: BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m (hunk 8 above)
Comment (danwalkowski) [inline BranchMonsterFactory/ViewControllers/MonsterCreatorViewController.m]: that was the original terrible name, not mine.
I'll change it

**Comment by danwalkowski** [comment]
ok all your issues cleared.

thumb?

File: BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m [hunk 18]
```diff
@@ -10,13 +10,18 @@
 #import "NetworkProgressBar.h"
 #import "MonsterViewerViewController.h"
 #import "MonsterPartsFactory.h"
-#import "MonsterPreferences.h"
 #import "Branch.h"
+#import "BranchUniversalObject.h"
+#import "BranchUniversalObject+MonsterHelpers.h"
 #import <MessageUI/MessageUI.h>
 #import <Social/Social.h>
 #import <FacebookSDK/FacebookSDK.h>
```
Comment (derrickstaten) [inline BranchMonsterFactory/ViewControllers/MonsterViewerViewController.m]: We're not using Facebook for anything, are we? Can we remove this completely? All imports and all files?

**Comment by derrickstaten** [comment]
@danwalkowski this looks really good now. Can you clean up the code (mostly spacing / new lines) now so that when others look at it, it shines? After you do that, :+1:

**Comment by derrickstaten** [comment]
- [ ] please confirm that Universal Links work for you on your physical device.

**Comment by danwalkowski** [comment]
Oh! regarding the External libs:  there was a bunch of third-party libraries for sharing.  The new method uses the official sharing libraries from the OS, so they aren't needed.

File: BranchMonsterFactory/Images.xcassets/LaunchImage.launchimage/Contents.json [hunk 19]
```diff
@@ -1,118 +0,0 @@
-{
```
Comment (derrickstaten) [inline BranchMonsterFactory/Images.xcassets/LaunchImage.launchimage/Contents.json]: why is this deleted?

Comment (derrickstaten) [inline BranchMonsterFactory/Images.xcassets/LaunchImage.launchimage/Contents.json]: i cloned a fresh copy, switched to this branch, and can't get it to run because this is missing

Comment (derrickstaten) [inline BranchMonsterFactory/Images.xcassets/LaunchImage.launchimage/Contents.json]:
```
CompileAssetCatalog /Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app BranchMonsterFactory/Images.xcassets BranchMonsterFactory/Media.xcassets
    cd /Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
    /Applications/Xcode.app/Contents/Developer/usr/bin/actool --output-format human-readable-text --notices --warnings --export-dependency-info /Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Intermediates/BranchMonsterFactory.build/Debug-iphoneos/BranchMonsterFactory.build/assetcatalog_dependencies --output-partial-info-plist /Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Intermediates/BranchMonsterFactory.build/Debug-iphoneos/BranchMonsterFactory.build/assetcatalog_generated_info.plist --app-icon AppIcon --launch-image LaunchImage --compress-pngs --enable-on-demand-resources YES --filter-for-device-model iPhone4,1 --filter-for-device-os-version 8.1 --target-device iphone --minimum-deployment-target 7.0 --platform iphoneos --compile /Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app /Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS/BranchMonsterFactory/Images.xcassets /Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS/BranchMonsterFactory/Media.xcassets

/* com.apple.actool.errors */
: error: None of the input catalogs contained a matching launch image set named  "LaunchImage".
/* com.apple.actool.notices */
: warning: Failed to read file attributes for "/Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS/BranchMonsterFactory/Media.xcassets"
    Failure Reason: No such file or directory
/* com.apple.actool.document.notices */
/Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS/BranchMonsterFactory/Images.xcassets:./AppIcon.appiconset/[iphone][57x57][][][1x]: notice: 57x57 app icons only apply to iPhone apps targeting releases of iOS prior to 7.0
/Users/derrickstaten/Developer/Branch-Example-Deep-Linking-Branchster-iOS/BranchMonsterFactory/Images.xcassets:./AppIcon.appiconset/[iphone][57x57][][][2x]: notice: 57x57@2x app icons only apply to iPhone apps targeting releases of iOS prior to 7.0
/* com.apple.actool.compilation-results */
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon29x29.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon29x29@2x.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon40x40@2x.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon57x57.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon57x57@2x.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon60x60@2x.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon29x29~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon29x29@2x~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon40x40~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon40x40@2x~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon50x50~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon50x50@2x~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon72x72~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon72x72@2x~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon76x76~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Products/Debug-iphoneos/Branchsters.app/AppIcon76x76@2x~ipad.png
/Users/derrickstaten/Library/Developer/Xcode/DerivedData/BranchMonsterFactory-bkrqdqnrrmwfdqfteeddfonptrxe/Build/Intermediates/BranchMonsterFactory.build/Debug-iphoneos/BranchMonsterFactory.build/assetcatalog_generated_info.plist
```

**Comment by Sarkar** [comment]
merging.
